File: FredBoat/src/main/java/fredboat/audio/player/GuildPlayer.java
/*
 * MIT License
 *
 * Copyright (c) 2017 Frederik Ar. Mikkelsen
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package fredboat.audio.player;

import fredboat.Config;
import fredboat.FredBoat;
import fredboat.audio.queue.AbstractTrackProvider;
import fredboat.audio.queue.AudioLoader;
import fredboat.audio.queue.AudioTrackContext;
import fredboat.audio.queue.IdentifierContext;
import fredboat.audio.queue.PersistentGuildTrackProvider;
import fredboat.audio.queue.RepeatMode;
import fredboat.audio.queue.SimpleTrackProvider;
import fredboat.commandmeta.MessagingException;
import fredboat.commandmeta.abs.CommandContext;
import fredboat.db.DatabaseNotReadyException;
import fredboat.db.EntityReader;
import fredboat.db.EntityWriter;
import fredboat.db.entity.common.AtcData;
import fredboat.db.entity.common.GuildConfig;
import fredboat.db.entity.common.GuildPlayerData;
import fredboat.feature.I18n;
import fredboat.feature.togglz.FeatureFlags;
import fredboat.messaging.CentralMessaging;
import fredboat.perms.PermissionLevel;
import fredboat.perms.PermsUtil;
import net.dv8tion.jda.core.JDA;
import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.TextChannel;
import net.dv8tion.jda.core.entities.VoiceChannel;
import net.dv8tion.jda.core.managers.AudioManager;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import java.io.IOException;
import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class GuildPlayer extends AbstractPlayer {

    private static final Logger log = LoggerFactory.getLogger(GuildPlayer.class);

    private final FredBoat shard;
    private final long guildId;
    public final Map<String, VideoSelection> selections = new HashMap<>(); //TODO possible source of leaks (holds audio tracks which aren't that lightweight)
    private long currentTCId;

    private final AudioLoader audioLoader;

    //dont save() while we are loading
    private boolean loading = true;

    @SuppressWarnings("LeakingThisInConstructor")
    public GuildPlayer(Guild guild) {
        super(guild.getId());
        log.debug("Constructing GuildPlayer({})", guild.getIdLong());

        onPlayHook = this::announceTrack;
        onErrorHook = this::handleError;

        this.shard = FredBoat.getInstance(guild.getJDA());
        this.guildId = guild.getIdLong();

        if (!LavalinkManager.ins.isEnabled()) {
            AudioManager manager = guild.getAudioManager();
            manager.setSendingHandler(this);
        }

        if (FeatureFlags.PERSISTENT_TRACK_PROVIDER.isActive()) {
            audioTrackProvider = new PersistentGuildTrackProvider(guild.getIdLong());
        } else {
            audioTrackProvider = new SimpleTrackProvider();
        }

        audioLoader = new AudioLoader(audioTrackProvider, getPlayerManager(), this);

        restoreGuildPlayer();
        loading = false;
    }

    /**
     * Recreates the state of the GuildPlayer from the database
     */
    private void restoreGuildPlayer() {
        log.debug("restoreGuildPlayer()");

        GuildPlayerData playerData;

        playerData = EntityReader.getEntity(guildId, GuildPlayerData.class);
        if (playerData == null) return;

        RepeatMode repeatMode = playerData.getRepeatMode();
        boolean shuffle = playerData.isShuffled();
        float volume = playerData.getVolume();

        this.setRepeatMode(repeatMode);
        if (Config.CONFIG.getDistribution().volumeSupported()) {
            this.setVolume(volume);
        }

        TextChannel tc = shard.getJda().getTextChannelById(playerData.getActiveTextChannelId());
        VoiceChannel vc = shard.getJda().getVoiceChannelById(playerData.getVoiceChannelId());
        if (tc != null) {
            this.currentTCId = tc.getIdLong();
        }
        boolean humansInVC = false;
        if (vc != null) {
            this.joinChannel(vc);
            humansInVC = !getHumanUsersInVC(vc).isEmpty();
        }

        AtcData atcData = EntityReader.getEntity(playerData.getPlayingTrackId(), AtcData.class);
        if (atcData != null) {
            try {
                AudioTrackContext trackContext = atcData.restoreTrack();
                audioTrackProvider.setLastTrack(trackContext);
                boolean silent = playerData.isPaused() || !humansInVC;
                playTrack(trackContext, silent);//dont announce the track if the player will be paused anyways
                trackContext.getTrack().setPosition(playerData.getPlayingTrackPosition());//this works just fine with live streams
            } catch (Exception ignored) {
            }
        }
        this.setShuffle(shuffle);

        //pause needs to be set after restoring the track, otherwise a different track may be loaded through the track provider
        if (!humansInVC) {
            this.setPause(true);
        } else {
            this.setPause(playerData.isPaused());
        }

        if (tc != null) {
            int trackCount = getTrackCount();
            if (trackCount > 0) {
                CentralMessaging.sendMessage(tc, MessageFormat.format(I18n.get(getGuild()).getString("reloadSuccess"), trackCount));
            }
        }
    }

    //call this whenever the state of this object changes
    //saving the track should only happen when you are planning to reconstruct this guild player again, for example when shutting down
    public void save(boolean... saveTrack) {
        if (loading) return;
        log.debug("GuildPlayer#save({}) called", saveTrack.length > 0 ? saveTrack[0] : "");

        Guild g = getGuild();
        if (g == null) return;

        Member self = g.getSelfMember();
        if (self == null) return;

        //NOTE: to save an updated voicechannel, save() needs to be called from the JDA event of us joining the
        // voicechannel, not from the request methods to do so
        VoiceChannel currentVC = self.getVoiceState().getChannel();
        log.debug("Current VoiceChannel is {}", currentVC == null ? "null" : currentVC.getIdLong());

        AudioTrackContext atc = null;
        if (saveTrack.length > 0 && saveTrack[0]) {
            atc = getPlayingTrack();
        }
        AtcData atcData = null;
        if (atc != null) try {
            atcData = new AtcData(atc);
        } catch (IOException ignored) {
        }

        GuildPlayerData gpd = new GuildPlayerData(guildId,
                currentVC != null ? currentVC.getIdLong() : 0,
                currentTCId,
                isPaused(),
                getVolume(),
                getRepeatMode(),
                isShuffle(),
                atcData != null ? atcData.getTrackId() : 0,
                atcData != null ? atc.getTrack().getPosition() : 0
        );

        GuildPlayerData old = EntityReader.getEntity(g.getIdLong(), GuildPlayerData.class);
        if (old == null || old.isDifferent(gpd)) {
            log.debug("Actually saving the guild player data due to differences");
            if (atcData != null) {
                EntityWriter.mergeAll(Arrays.asList(gpd, atcData));
            } else {
                EntityWriter.merge(gpd);
            }
        }
    }

    private void announceTrack(AudioTrackContext atc) {
        if (getRepeatMode() != RepeatMode.SINGLE && isTrackAnnounceEnabled() && !isPaused()) {
            TextChannel activeTextChannel = getActiveTextChannel();
            if (activeTextChannel != null) {
                CentralMessaging.sendMessage(activeTextChannel,
                        MessageFormat.format(I18n.get(getGuild()).getString("trackAnnounce"), atc.getEffectiveTitle()));
            }
        }
    }

    private void handleError(Throwable t) {
        if (!(t instanceof MessagingException)) {
            log.error("Guild player error", t);
        }
        TextChannel tc = getActiveTextChannel();
        if (tc != null) {
            CentralMessaging.sendMessage(tc, "Something went wrong!\n" + t.getMessage());
        }
    }

    public void joinChannel(Member usr) throws MessagingException {
        VoiceChannel targetChannel = getUserCurrentVoiceChannel(usr);
        joinChannel(targetChannel);
    }

    public void joinChannel(VoiceChannel targetChannel) throws MessagingException {
        if (targetChannel == null) {
            throw new MessagingException(I18n.get(getGuild()).getString("playerUserNotInChannel"));
        }
        if (targetChannel.equals(getCurrentVoiceChannel(targetChannel.getJDA()))) {
            // already connected to the channel
            return;
        }

        if (!targetChannel.getGuild().getSelfMember().hasPermission(targetChannel, Permission.VOICE_CONNECT)
                && !targetChannel.getMembers().contains(getGuild().getSelfMember())) {
            throw new MessagingException(I18n.get(getGuild()).getString("playerJoinConnectDenied"));
        }

        if (!targetChannel.getGuild().getSelfMember().hasPermission(targetChannel, Permission.VOICE_SPEAK)) {
            throw new MessagingException(I18n.get(getGuild()).getString("playerJoinSpeakDenied"));
        }

        LavalinkManager.ins.openConnection(targetChannel);
        AudioManager manager = getGuild().getAudioManager();
        manager.setConnectionListener(new DebugConnectionListener(guildId, shard.getShardInfo()));

        log.info("Connected to voice channel " + targetChannel);
    }

    public void leaveVoiceChannelRequest(CommandContext commandContext, boolean silent) {
        if (!silent) {
            VoiceChannel currentVc = LavalinkManager.ins.getConnectedChannel(commandContext.guild);
            if (currentVc == null) {
                commandContext.reply(I18n.get(getGuild()).getString("playerNotInChannel"));
            } else {
                commandContext.reply(MessageFormat.format(I18n.get(getGuild()).getString("playerLeftChannel"),
                        currentVc.getName()));
            }
        }
        LavalinkManager.ins.closeConnection(getGuild());
    }

    /**
     * May return null if the member is currently not in a channel
     */
    @Nullable
    public VoiceChannel getUserCurrentVoiceChannel(Member member) {
        return member.getVoiceState().getChannel();
    }

    public void queue(String identifier, CommandContext context) {
        IdentifierContext ic = new IdentifierContext(identifier, context.channel, context.invoker);

        if (context.invoker != null) {
            joinChannel(context.invoker);
        }

        audioLoader.loadAsync(ic);
    }

    public void queue(IdentifierContext ic) {
        if (ic.getMember() != null) {
            joinChannel(ic.getMember());
        }

        audioLoader.loadAsync(ic);
    }

    public void queue(AudioTrackContext atc){
        Member member = getGuild().getMemberById(atc.getUserId());
        if (member != null) {
            joinChannel(member);
        }
        audioTrackProvider.add(atc);
        play();
    }

    public int getTrackCount() {
        int trackCount = audioTrackProvider.size();
        if (player.getPlayingTrack() != null) trackCount++;
        return trackCount;
    }

    public List<AudioTrackContext> getTracksInRange(int start, int end) {
        log.debug("getTracksInRange({} {})", start, end);

        List<AudioTrackContext> result = new ArrayList<>();

        //adjust args for whether there is a track playing or not
        if (player.getPlayingTrack() != null) {
            if (start <= 0) {
                result.add(context);
                end--;//shorten the requested range by 1, but still start at 0, since that's the way the trackprovider counts its tracks
            } else {
                //dont add the currently playing track, drop the args by one since the "first" track is currently playing
                start--;
                end--;
            }
        } else {
            //nothing to do here, args are fine to pass on
        }

        result.addAll(audioTrackProvider.getTracksInRange(start, end));
        return result;
    }

    //similar to getTracksInRange, but gets the trackIds only
    public List<Long> getTrackIdsInRange(int start, int end) {
        log.debug("getTrackIdsInRange({} {})", start, end);

        List<Long> result = new ArrayList<>();

        //adjust args for whether there is a track playing or not
        if (player.getPlayingTrack() != null) {
            if (start <= 0) {
                result.add(context.getTrackId());
                end--;//shorten the requested range by 1, but still start at 0, since that's the way the trackprovider counts its tracks
            } else {
                //dont add the currently playing track, drop the args by one since the "first" track is currently playing
                start--;
                end--;
            }
        } else {
            //nothing to do here, args are fine to pass on
        }

        // optimization for the persistent track provider
        if (audioTrackProvider instanceof PersistentGuildTrackProvider) {
            PersistentGuildTrackProvider trackProvider = (PersistentGuildTrackProvider) audioTrackProvider;
            result.addAll(trackProvider.getTrackIdsInRange(start, end));
        } else {
            result.addAll(audioTrackProvider.getTracksInRange(start, end).stream()
                    .map(AudioTrackContext::getTrackId)
                    .collect(Collectors.toList()));
        }

        return result;
    }

    public long getTotalRemainingMusicTimeMillis() {
        //Live streams are considered to have a length of 0
        long millis = audioTrackProvider.getDurationMillis();

        AudioTrackContext currentTrack = player.getPlayingTrack() != null ? context : null;
        if (currentTrack != null && !currentTrack.getTrack().getInfo().isStream) {
            millis += Math.max(0, currentTrack.getEffectiveDuration() - getPosition());
        }
        return millis;
    }


    public long getStreamsCount() {
        long streams = audioTrackProvider.streamsCount();
        AudioTrackContext atc = player.getPlayingTrack() != null ? context : null;
        if (atc != null && atc.getTrack().getInfo().isStream) streams++;
        return streams;
    }


    //optionally pass a jda object to use for the lookup
    @Nullable
    public VoiceChannel getCurrentVoiceChannel(JDA... jda) {
        JDA j;
        if (jda.length == 0) {
            j = getJda();
        } else {
            j = jda[0];
        }
        Guild guild = j.getGuildById(guildId);
        if (guild != null)
            return LavalinkManager.ins.getConnectedChannel(guild);
        else
            return null;
    }

    /**
     * @return the text channel currently used for music commands, if there is none return the default channel
     */
    @Nullable
    public TextChannel getActiveTextChannel() {
        TextChannel currentTc = getCurrentTC();
        if (currentTc != null) {
            return currentTc;
        } else {
            log.warn("No currentTC in " + getGuild() + "! Returning default channel...");
            return getGuild().getDefaultChannel();
        }

    }

    @Nonnull
    public List<Member> getHumanUsersInVC(@Nullable VoiceChannel vc) {
        if (vc == null) {
            return Collections.emptyList();
        }

        ArrayList<Member> nonBots = new ArrayList<>();
        for (Member member : vc.getMembers()) {
            if (!member.getUser().isBot()) {
                nonBots.add(member);
            }
        }
        return nonBots;
    }

    /**
     * @return Users who are not bots
     */
    public List<Member> getHumanUsersInCurrentVC() {
        return getHumanUsersInVC(getCurrentVoiceChannel());
    }

    @Override
    public String toString() {
        return "[GP:" + getGuild().getId() + "]";
    }

    public Guild getGuild() {
        return getJda().getGuildById(guildId);
    }

    public RepeatMode getRepeatMode() {
        if (audioTrackProvider instanceof AbstractTrackProvider)
            return ((AbstractTrackProvider) audioTrackProvider).getRepeatMode();
        else return RepeatMode.OFF;
    }

    public boolean isShuffle() {
        return audioTrackProvider instanceof AbstractTrackProvider && ((AbstractTrackProvider) audioTrackProvider).isShuffle();
    }

    public void setRepeatMode(RepeatMode repeatMode) {
        if (audioTrackProvider instanceof AbstractTrackProvider) {
            ((AbstractTrackProvider) audioTrackProvider).setRepeatMode(repeatMode);
        } else {
            throw new UnsupportedOperationException("Can't repeat " + audioTrackProvider.getClass());
        }
        save();
    }

    public void setShuffle(boolean shuffle) {
        if (audioTrackProvider instanceof AbstractTrackProvider) {
            ((AbstractTrackProvider) audioTrackProvider).setShuffle(shuffle);
        } else {
            throw new UnsupportedOperationException("Can't shuffle " + audioTrackProvider.getClass());
        }
        save();
    }

    public void reshuffle() {
        if (audioTrackProvider instanceof AbstractTrackProvider) {
            ((AbstractTrackProvider) audioTrackProvider).reshuffle();
        } else {
            throw new UnsupportedOperationException("Can't reshuffle " + audioTrackProvider.getClass());
        }
    }

    public void setCurrentTC(TextChannel tc) {
        if (this.currentTCId != tc.getIdLong()) {
            this.currentTCId = tc.getIdLong();
            save();
        }

    }

    @Override
    public void setPause(boolean pause) {
        super.setPause(pause);
        save();
    }

    @Override
    public void setVolume(float vol) {
        super.setVolume(vol);
        save();
    }

    /**
     * @return currently used TextChannel or null if there is none
     */
    private TextChannel getCurrentTC() {
        try {
            return shard.getJda().getTextChannelById(currentTCId);
        } catch (IllegalArgumentException e) {
            return null;
        }
    }

    //Success, fail message
    public Pair<Boolean, String> canMemberSkipTracks(Member member, Collection<Long> trackIds) {
        if (PermsUtil.checkPerms(PermissionLevel.DJ, member)) {
            return new ImmutablePair<>(true, null);
        } else {
            //We are not a mod
            long userId = member.getUser().getIdLong();

            //if there is a currently playing track, and the track is requested to be skipped, but not owned by the
            // requesting user, then currentTrackSkippable should be false
            boolean currentTrackSkippable = true;
            AudioTrackContext playingTrack = getPlayingTrack();
            if (playingTrack != null
                    && trackIds.contains(getPlayingTrack().getTrackId())
                    && playingTrack.getUserId() != userId) {

                currentTrackSkippable = false;
            }

            if (currentTrackSkippable
                    && audioTrackProvider.isUserTrackOwner(userId, trackIds)) { //check ownership of the queued tracks
                return new ImmutablePair<>(true, null);
            } else {
                return new ImmutablePair<>(false, I18n.get(getGuild()).getString("skipDeniedTooManyTracks"));
            }
        }
    }

    public void skipTracksForMemberPerms(CommandContext context, Collection<Long> trackIds, String successMessage) {
        Pair<Boolean, String> pair = canMemberSkipTracks(context.invoker, trackIds);

        if (pair.getLeft()) {
            context.reply(successMessage);
            skipTracks(trackIds);
        } else {
            context.replyWithName(pair.getRight());
        }
    }

    private void skipTracks(Collection<Long> trackIds) {
        boolean skipCurrentTrack = false;

        List<Long> toRemove = new ArrayList<>();
        AudioTrackContext playing = player.getPlayingTrack() != null ? context : null;
        for (Long trackId : trackIds) {
            if (playing != null && trackId.equals(playing.getTrackId())) {
                //Should be skipped last, in respect to PlayerEventListener
                skipCurrentTrack = true;
            } else {
                toRemove.add(trackId);
            }
        }

        audioTrackProvider.removeAllById(toRemove);

        if (skipCurrentTrack) {
            skip();
        }
    }

    private boolean isTrackAnnounceEnabled() {
        boolean enabled = false;
        try {
            GuildConfig config = EntityReader.getOrCreateEntity(Long.toString(guildId), GuildConfig.class);
            enabled = config.isTrackAnnounce();
        } catch (DatabaseNotReadyException ignored) {}

        return enabled;
    }

    public JDA getJda() {
        return shard.getJda();
    }

    @Override
    void destroy() {
        EntityWriter.deleteObject(guildId, GuildPlayerData.class);
        super.destroy();
        log.info("Player for " + guildId + " was destroyed.");
    }
}

File: FredBoat/src/main/java/fredboat/audio/player/GuildPlayer.kt
/*
 * MIT License
 *
 * Copyright (c) 2017 Frederik Ar. Mikkelsen
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package fredboat.audio.player

import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import com.sedmelluq.discord.lavaplayer.track.TrackMarker
import fredboat.audio.lavalink.SentinelLavalink
import fredboat.audio.lavalink.SentinelLink
import fredboat.audio.queue.*
import fredboat.audio.queue.limiter.QueueLimitStatus
import fredboat.audio.queue.limiter.QueueLimiter
import fredboat.audio.queue.handlers.*
import fredboat.command.music.control.VoteSkipCommand
import fredboat.commandmeta.MessagingException
import fredboat.commandmeta.abs.CommandContext
import fredboat.db.api.GuildSettingsRepository
import fredboat.definitions.RepeatMode
import fredboat.sentinel.Guild
import fredboat.sentinel.InternalGuild
import fredboat.sentinel.TextChannel
import fredboat.util.TextUtils
import fredboat.util.extension.escapeAndDefuse
import fredboat.util.ratelimit.Ratelimiter
import fredboat.util.rest.YoutubeAPI
import fredboat.ws.emptyPlayerInfo
import fredboat.ws.toPlayerInfo
import lavalink.client.player.IPlayer
import lavalink.client.player.LavalinkPlayer
import lavalink.client.player.event.PlayerEventListenerAdapter
import org.bson.types.ObjectId
import org.json.JSONObject
import org.slf4j.LoggerFactory
import reactor.core.publisher.Mono
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.function.Consumer
import javax.annotation.CheckReturnValue

class GuildPlayer(
        val lavalink: SentinelLavalink,
        var guild: Guild,
        private val musicTextChannelProvider: MusicTextChannelProvider,
        audioPlayerManager: AudioPlayerManager,
        private val guildSettingsRepository: GuildSettingsRepository,
        ratelimiter: Ratelimiter,
        youtubeAPI: YoutubeAPI
) : PlayerEventListenerAdapter() {

    val queueHandler: IQueueHandler = RepeatableQueueHandler(this)
    private val audioLoader = AudioLoader(ratelimiter, queueHandler, audioPlayerManager, this, youtubeAPI)
    private val queueLimiter = QueueLimiter(guildSettingsRepository)
    val guildId = guild.id
    val player: LavalinkPlayer = lavalink.getLink(guild.id.toString()).player
    var internalContext: AudioTrackContext? = null

    private var onPlayHook: Consumer<AudioTrackContext>? = null
    private var onErrorHook: Consumer<Throwable>? = null
    @Volatile
    private var lastLoadedTrack: AudioTrackContext? = null
    val historyQueue = ConcurrentLinkedQueue<AudioTrackContext>()

    companion object {
        private val log = LoggerFactory.getLogger(GuildPlayer::class.java)
        private const val MAX_HISTORY_SIZE = 20
    }

    /**
     * @return The text channel currently used for music commands.
     *
     * May return null if the channel was deleted.
     */
    val activeTextChannel: TextChannel?
        get() {
            if (!guild.selfPresent) return null
            return musicTextChannelProvider.getMusicTextChannel(guild)
        }

    var repeatMode: RepeatMode
        get() = if (queueHandler is IRepeatableQueueHandler)
            queueHandler.repeat
        else
            RepeatMode.OFF
        set(repeatMode) = if (queueHandler is IRepeatableQueueHandler) {
            queueHandler.repeat = repeatMode
        } else {
            throw UnsupportedOperationException("Can't repeat " + queueHandler.javaClass)
        }

    var isRoundRobin : Boolean
        get() = queueHandler is IRoundRobinQueueHandler && queueHandler.roundRobin
        set(value) = if (queueHandler is IRoundRobinQueueHandler) {
            queueHandler.roundRobin = value
        } else {
            throw UnsupportedOperationException("Can't round robin " + queueHandler.javaClass)
        }

    var isShuffle: Boolean
        get() = queueHandler is IShufflableQueueHandler && queueHandler.shuffle
        set(shuffle) = if (queueHandler is IShufflableQueueHandler) {
            queueHandler.shuffle = shuffle
            internalContext?.isPriority = false
        } else {
            throw UnsupportedOperationException("Can't shuffle " + queueHandler.javaClass)
        }

    private fun getTrackAnnounceStatus(): Mono<Boolean> {
            if (guild.selfPresent) {
                 return guildSettingsRepository.fetch(guild.id).map { it.trackAnnounce }
            }

            return Mono.just(false)
        }

    val playingTrack: AudioTrackContext?
        get() {
            return if (player.playingTrack == null && internalContext == null) {
                queueHandler.peek()
            } else internalContext
        }

    //the unshuffled playlist
    //Includes currently playing track, which comes first
    val remainingTracks: List<AudioTrackContext>
        get() {
            log.trace("getRemainingTracks()")
            val list = ArrayList<AudioTrackContext>()
            val atc = playingTrack
            if (atc != null) {
                list.add(atc)
            }

            list.addAll(queueHandler.queue.toList())
            return list
        }

    var volume: Float
        get() = player.volume.toFloat() / 100
        set(vol) {
            player.volume = (vol * 100).toInt()
        }

    val isPlaying: Boolean
        get() = player.playingTrack != null && !player.isPaused

    val isPaused: Boolean
        get() = player.isPaused

    val position: Long
        get() = player.trackPosition

    init {
        log.debug("Constructing GuildPlayer({})", guild)
        onPlayHook = Consumer { this.announceTrack(it) }
        onErrorHook = Consumer { this.handleError(it) }
        @Suppress("LeakingThis")
        player.addListener(this)
        linkPostProcess()
    }

    private fun announceTrack(atc: AudioTrackContext) {
        val activeTextChannel = activeTextChannel

        getTrackAnnounceStatus().subscribe {
            if (it && !isPaused && repeatMode != RepeatMode.SINGLE) {
                activeTextChannel?.send(atc.i18nFormat(
                        "trackAnnounce",
                        atc.effectiveTitle.escapeAndDefuse(),
                        atc.member.effectiveName.escapeAndDefuse()
                ))?.subscribe()
            }
        }
    }

    private fun handleError(t: Throwable) {
        if (t !is MessagingException) {
            log.error("Guild player error", t)
        }
        val activeTextChannel = activeTextChannel
        activeTextChannel?.send("Something went wrong!\n${t.message}")?.subscribe()
    }

    fun queueAsync(identifier: String, context: CommandContext, isPriority: Boolean = false) {
        val ic = IdentifierContext(identifier, context.textChannel, context.member)
        ic.isPriority = isPriority

        joinChannel(context.member)

        audioLoader.loadAsync(ic)
    }

    fun queueAsync(ic: IdentifierContext) {
        joinChannel(ic.member)

        audioLoader.loadAsync(ic)
    }

    fun queue(atc: AudioTrackContext) {
        if (!guild.selfPresent) throw IllegalStateException("Attempt to queue track in a guild we are not present in")

        val member = guild.getMember(atc.userId)
        if (member != null) {
            joinChannel(member)
        }

        queueHandler.add(atc)
        if (isPlaying) updateClients()
    }

    /** Add a bunch of tracks to the track provider */
    fun queueAll(tracks: Collection<AudioTrackContext>) {
        queueHandler.addAll(tracks)
    }

    @CheckReturnValue
    suspend fun queueLimited(atc: AudioTrackContext): QueueLimitStatus {
        val status = queueLimiter.isQueueLimited(atc, this)

        // only queue if track was not limited
        if (status.canQueue) {
            queue(atc)
        }

        return status
    }

    suspend fun queueLimited(tracks: List<AudioPlaylistContext>): List<QueueLimitStatus> {
        val states = queueLimiter.isQueueLimited(tracks, this)

        queueAll(states.filter { it.canQueue }.map { it.atc })
        return states
    }


    override fun toString(): String {
        return "[GP:$guildId]"
    }

    fun reshuffle() {
        if (queueHandler is IShufflableQueueHandler) {
            queueHandler.reshuffle()
            internalContext?.isPriority = false
            updateClients()
        } else {
            throw UnsupportedOperationException("Can't reshuffle " + queueHandler.javaClass)
        }
    }

    fun skipTracks(trackIds: Collection<ObjectId>) {
        var skipCurrentTrack = false

        val toRemove = ArrayList<ObjectId>()
        val playing = if (player.playingTrack != null) internalContext else null
        for (trackId in trackIds) {
            if (playing != null && trackId == playing.trackId) {
                //Should be skipped last, in respect to PlayerEventListener
                skipCurrentTrack = true
            } else {
                toRemove.add(trackId)
            }
        }

        if (toRemove.size > 0) queueHandler.removeById(toRemove)
        if (skipCurrentTrack) skip()
    }

    override fun onTrackStart(player: IPlayer?, track: AudioTrack?) {
        voteSkipCleanup()
        super.onTrackStart(player, track)
    }

    fun destroy() {
        queueHandler.clear()
        stop()
        player.removeListener(this)
        player.link.destroy()
        log.info("Player for $guildId was destroyed.")
        lavalink.userSessionHandler.sendLazy(guildId) { emptyPlayerInfo }
    }

    private fun voteSkipCleanup() {
        VoteSkipCommand.guildSkipVotes.remove(guildId)
    }

    /**
     * Invoked when subscribing to this player's guild, with an already existing guild
     */
    fun linkPostProcess() {
        val iGuild = guild as InternalGuild
        val vsu = iGuild.cachedVsu
        val slink = player.link as SentinelLink
        if (vsu != null) {
            iGuild.cachedVsu = null
            slink.onVoiceServerUpdate(JSONObject(vsu.raw), vsu.sessionId)
            log.info("Using cached VOICE_SERVER_UPDATE for $guild")

            val vc = guild.selfMember.voiceChannel
            if (vc == null)
                log.warn("Using cached VOICE_SERVER_UPDATE, but it doesn't appear like we are in a voice channel!")
            else
                slink.setChannel(vc.idString)
        } else {
            val vc = slink.getChannel(guild) ?: return
            slink.connect(vc, skipIfSameChannel = false)
        }
        updateClients()
    }

    fun play() {
        log.trace("play()")

        if (player.isPaused) {
            player.isPaused = false
        }
        if (player.playingTrack == null) {
            logListeners()
            loadAndPlay()
        }

    }

    fun setPause(pause: Boolean) {
        log.trace("setPause({})", pause)

        if (pause) {
            player.isPaused = true
        } else {
            player.isPaused = false
            play()
        }
        updateClients()
    }

    fun pause() = setPause(true)

    /**
     * Clear the tracklist and stop the current track
     */
    fun stop() {
        log.trace("stop()")

        queueHandler.clear()
        stopTrack()
    }

    /**
     * Skip the current track
     */
    fun skip() {
        log.trace("skip()")

        queueHandler.onSkipped()
        stopTrack()
    }

    /**
     * Stop the current track.
     */
    fun stopTrack() {
        log.trace("stopTrack()")

        internalContext = null
        player.stopTrack()
    }

    override fun onTrackEnd(player: IPlayer?, track: AudioTrack?, endReason: AudioTrackEndReason?) {
        log.debug("onTrackEnd({} {} {}) called", track!!.info.title, endReason!!.name, endReason.mayStartNext)

        if (endReason == AudioTrackEndReason.FINISHED || endReason == AudioTrackEndReason.STOPPED) {
            updateHistoryQueue()
            loadAndPlay()
        } else if (endReason == AudioTrackEndReason.CLEANUP) {
            log.info("Track " + track.identifier + " was cleaned up")
        } else if (endReason == AudioTrackEndReason.LOAD_FAILED) {
            if (onErrorHook != null)
                onErrorHook!!.accept(MessagingException("Track `" + TextUtils.escapeAndDefuse(track.info.title) + "` failed to load. Skipping..."))
            queueHandler.onSkipped()
            loadAndPlay()
        } else {
            log.warn("Track " + track.identifier + " ended with unexpected reason: " + endReason)
        }
    }

    //request the next track from the track provider and start playing it
    private fun loadAndPlay() {
        log.trace("loadAndPlay()")

        val atc = queueHandler.take()
        lastLoadedTrack = atc
        atc?.let { playTrack(it) }
        updateClients()
    }

    private fun updateHistoryQueue() {
        val lastTrack = lastLoadedTrack
        if (lastTrack == null) {
            log.warn("No lastLoadedTrack in $this after track end")
            return
        }
        if (historyQueue.size == MAX_HISTORY_SIZE) {
            historyQueue.poll()
        }
        historyQueue.add(lastTrack)
    }

    /**
     * Plays the provided track.
     *
     * Silently playing a track will not trigger the onPlayHook (which announces the track usually)
     */
    private fun playTrack(trackContext: AudioTrackContext, silent: Boolean = false) {
        log.trace("playTrack({})", trackContext.effectiveTitle)

        internalContext = trackContext
        player.playTrack(trackContext.track)
        trackContext.track.position = trackContext.startPosition

        if (trackContext is SplitAudioTrackContext) {
            //Ensure we don't step over our bounds
            log.info("Start: ${trackContext.startPosition} End: ${trackContext.startPosition + trackContext.effectiveDuration}")

            trackContext.track.setMarker(
                    TrackMarker(trackContext.startPosition + trackContext.effectiveDuration,
                            TrackEndMarkerHandler(this, trackContext)))
        }

        if (!silent && onPlayHook != null) onPlayHook!!.accept(trackContext)
    }

    override fun onTrackException(player: IPlayer?, track: AudioTrack, exception: Exception?) {
        log.error("Lavaplayer encountered an exception while playing {}",
                track.identifier, exception)
    }

    override fun onTrackStuck(player: IPlayer?, track: AudioTrack, thresholdMs: Long) {
        log.error("Lavaplayer got stuck while playing {}",
                track.identifier)
    }

    fun seekTo(position: Long) {
        if (internalContext!!.track.isSeekable) {
            player.seekTo(position)
            updateClients()
        } else {
            throw MessagingException(internalContext!!.i18n("seekDeniedLiveTrack"))
        }
    }

    private fun logListeners() {
        humanUsersInCurrentVC.forEach { lavalink.activityMetrics.logListener(it) }
    }

    private fun updateClients() {
        lavalink.userSessionHandler.sendLazy(guildId) { toPlayerInfo() }
    }
}
